### PR TITLE
MuE distributions for Pyro.

### PR DESCRIPTION
**Pull request by EWeinstein**
An implementation of MuE models for pyro. The code is organized as:

1. `contrib/mue/statearrangers.py` Modules for converting standard MuE parameters into standard HMM parameters. Currently, this is just implemented for the profile HMM architecture, but in the future other MuE models will be added.
2. `contrib/mue/variablelengthhmm.py` An HMM with discrete states and discrete emissions which can handle variable length sequences, assuming that they are one hot encoded and padded at the end with zeros to some fixed length. Based on DiscreteHMM. In the future this will be extended with functions for sampling, filtering, estimating the latent state and computing the first moment of the distribution. 
3. `contrib/mue/biosequenceloaders.py` Currently empty; in the future, this will include torch dataloaders for fasta files and for a2m (MSA) files.

Example models (profile HMM, FactorMuE) can be found in `examples/contrib/mue`. Unit tests can be found in `tests/contrib/mue`.

One more note: Pyro's DiscreteHMM computes the probability of the initial latent state as `initial_probs @ transition_probs` rather than just using `initial_probs` (the more common convention I think). I used the second convention, but if Pyro is committed to the first I will change my HMM appropriately; it might help to clarify this in the documentation. (I found the same behavior in Edward2, https://github.com/tensorflow/probability/issues/958).

## Tasks

- [x] Add a file docs/source/contrib.mue.rst similar to other contrib.*.rst files. Run `make docs` and view docs/build/html/contrib.mue.html to ensure this works.
- [x] Add files tutorial/source/*.rst for each example, following other .rst files by including code. Run `make tutorial` and examine docs to ensure this works.
- [x] Add shape tests for `VariableLengthDiscreteHMM`
- [x] Add tests comparing `VariableLengthDiscreteHMM` vs `DiscreteHMM`
- [x] Add unit tests for `Profile`
- [x] Register both examples in `CPU_EXAMPLES` of [test_examples.py](https://github.com/pyro-ppl/pyro/blob/dev/tests/test_examples.py). We try to pass args so that each such test is quick to run. Run `make test-examples` to ensure this works.

### CODE REVIEW

**Comment by eb8680** [comment]
cc @samuelklee

File: tests/contrib/mue/test_variablelengthhmm.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import torch
+
+from pyro.contrib.mue.variablelengthhmm import VariableLengthDiscreteHMM
+
+
+def test_hmm_log_prob():
+    torch.set_default_tensor_type('torch.DoubleTensor')
```
Comment (fritzo) [inline tests/contrib/mue/test_variablelengthhmm.py:7]: nit: This should happen automatically due to [conftest.py](https://github.com/pyro-ppl/pyro/blob/dev/tests/conftest.py#L13). We try to avoid setting global variables in individual tests because those can leak across tests.

File: examples/contrib/mue/phmm.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    plt.xlabel('position')
+    plt.ylabel('probability of delete')
+    plt.savefig('phmm_plot.delete_prob_{}.pdf'.format(time_stamp))
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description="Basic profile HMM model (constant + MuE).")
+    parser.add_argument('-t', '--test', action='store_true', default=False,
+                        help='small dataset, a few steps')
```
Comment (fritzo) [inline examples/contrib/mue/phmm.py:197]: Great! We can add a line to `CPU_EXAMPLES` in [test_examples.py](https://github.com/pyro-ppl/pyro/blob/dev/tests/test_examples.py#L18):
```py
'contrib/mue/phmm.py --test',
```

**Review by fritzo — COMMENTED** [review]
Looks good so far. Let's chat later this week. It's probably worth adding a tutorial notebook in a subsequent PR.

**Comment by fritzo** [comment]
Another thing @eb8680 and @martinjankowiak and I have discussed is creating a new `biopyro` package that moves all biology-related stuff out of Pyro. That would be a good thing to do in the long term, say before Pyro 2.0; in the short term it seems good to put MuE in `pyro.contrib`. We might consider making a `pyro.contrib.dataloaders` for fasta files etc, since they are useful for other genetics models.

File: examples/contrib/mue/FactorMuE.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+    def forward(self, z):
+
+        seq = self.f(z)
+        seq = seq.reshape([-1, 2, self.latent_seq_length+1,
+                           self.alphabet_length])
+        return seq
+
+
+class FactorMuE(nn.Module):
```
Comment (fritzo) [inline examples/contrib/mue/FactorMuE.py:62]: As mentioned offline, consider moving this to something like `pyro/contrib/mue/models/factormue.py`, where `.../models` contains somewhat-reusable example models. I like how that has worked out in [pyro.contrib.epidemiology.models](http://docs.pyro.ai/en/stable/_modules/pyro/contrib/epidemiology/models.html).

File: pyro/contrib/mue/statearrangers.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import torch
+import torch.nn as nn
+
+
+def mg2k(m, g):
+    """Convert from (m, g) indexing to k indexing."""
+    return 2*m + 1 - g
+
+
+class profile(nn.Module):
```
Comment (fritzo) [inline pyro/contrib/mue/statearrangers.py:12]: nit: class names should be upper case

File: pyro/contrib/mue/statearrangers.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
```
Comment (fritzo) [inline pyro/contrib/mue/statearrangers.py:13]: Add a docstring, including:
- meaning, type, and shape of each input parameter (see other Pyro classes for example sphinx notation)
- space and time complexity using `O(-)` notation, and
- one or two citations.

File: pyro/contrib/mue/__init__.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+"""
+The :mod:`pyro.contrib.mue` module provides tools for working with mutational
+emission (MuE) distributions.
+See Weinstein and Marks (2020),
+https://www.biorxiv.org/content/10.1101/2020.07.31.231381v1.
+Primary developer is Eli N. Weinstein (https://eweinstein.github.io/).
+"""
```
Comment (fritzo) [inline pyro/contrib/mue/__init__.py:9]: Note you can either keep this text here, or move it to docs/source/contrib.mue.rst. From that .rst file you should use `automodule` to include this module. See other .rst files there for example.

File: pyro/contrib/mue/variablelengthhmm.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+# SPDX-License-Identifier: Apache-2.0
+import torch
+from torch.distributions import constraints
+
+from pyro.distributions.hmm import _sequential_logmatmulexp
+from pyro.distributions.torch_distribution import TorchDistribution
+from pyro.distributions.util import broadcast_shape
+
+
+class VariableLengthDiscreteHMM(TorchDistribution):
```
Comment (fritzo) [inline pyro/contrib/mue/variablelengthhmm.py:11]: This is a really cool distribution 👍 I'd consider renaming it to maybe `MissingDataDiscreteHMM` or `OneHotDiscreteHMM` since it is quite a bit more general than `VariableLength`. Once it settles down, we could even move it to `pyro.distributions`.

File: pyro/contrib/mue/variablelengthhmm.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    """
+    HMM with discrete latent states and discrete observations, allowing for
+    variable length sequences.
+    """
```
Comment (fritzo) [inline pyro/contrib/mue/variablelengthhmm.py:15]: describe all input parameters.

File: pyro/contrib/mue/variablelengthhmm.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        super(VariableLengthDiscreteHMM, self).__init__(
+            batch_shape, event_shape, validate_args=validate_args)
+
+    def log_prob(self, value):
+        """Warning: unlike in pyro's DiscreteHMM, which computes the
+        probability of the first state as
+        initial.T @ transition @ emission
+        this distribution uses the standard HMM convention,
+        initial.T @ emission
+        """
```
Comment (fritzo) [inline pyro/contrib/mue/variablelengthhmm.py:54]: Move this into the class docstring and use sphinx warning notation:
```
.. warning: Unlike in Pyro's :class:`~pyro.distributions.DiscreteHMM`, which
    computes the probability of the first state as ``initial.T @ transition @ emission``,
    this distribution uses the standard HMM convention
    ``initial.T @ emission``.
```

File: tests/contrib/mue/test_statearrangers.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+from pyro.contrib.mue.statearrangers import mg2k, profile
+
+
+def simpleprod(lst):
+    # Product of list of scalar tensors, as numpy would do it.
+    if len(lst) == 0:
+        return torch.tensor(1.)
+    else:
+        return torch.prod(torch.cat([elem[None] for elem in lst]))
```
Comment (fritzo) [inline tests/contrib/mue/test_statearrangers.py:12]: nit: you could simplify to `torch.stack(lst).prod()`

File: tests/contrib/mue/test_statearrangers.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        ll = torch.rand((4, 5))
+        ll = ll/torch.sum(ll, dim=1, keepdim=True)
+        a0ln, aln, eln = pf_arranger.forward(torch.log(s), torch.log(c),
+                                             torch.log(r), torch.log(u),
+                                             torch.log(ll))
+    else:
+        a0ln, aln, eln = pf_arranger.forward(torch.log(s), torch.log(c),
+                                             torch.log(r), torch.log(u))
+
+    # - Remake transition matrices. -
```
Comment (fritzo) [inline tests/contrib/mue/test_statearrangers.py:48]: Could you add a little bit of commenting to this? Just like what each big code block is doing? And maybe a reference equation?

Also as we discussed, you might see if there is a simpler test for some special cases such as when indel prob goes to zero.

File: pyro/contrib/mue/biosequenceloaders.py [hunk 12]
```diff
@@ -0,0 +1,2 @@
+# Copyright Contributors to the Pyro project.
```
Comment (fritzo) [inline pyro/contrib/mue/biosequenceloaders.py:1]: nit: maybe name this file `dataloaders`?

**Review by fritzo — COMMENTED** [review]
Looks great!

File: pyro/contrib/mue/missingdatahmm.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        to ``(batch_size, state_dim, state_dim)``.
+    :param ~torch.Tensor observation_logits: A logits tensor for observation
+        distributions from latent states. Should have rightmost shape
+        ``(state_dim, categorical_size)``, where ``categorical_size`` is the
+        dimension of the categorical output, and be broadcastable
+        to ``(batch_size, state_dim, categorical_size)``.
+    """
+    arg_constraints = {"initial_logits": constraints.real,
+                       "transition_logits": constraints.real,
+                       "observation_logits": constraints.real}
```
Comment (fritzo) [inline pyro/contrib/mue/missingdatahmm.py:39]: These will need to be more precise:
```py
arg_constraints = {"initial_logits": constraints.real_vector,
                   "transition_logits": constraints.independent(constraints.real, 2),
                   "observation_logits": constraints.independent(constraints.real, 2)}
support = constraints.independent(constraints.nonnegative_integer, 2)
```

File: tests/contrib/mue/test_seqs.fasta [hunk 14]
```diff
@@ -0,0 +1,7 @@
+>one
```
Comment (fritzo) [inline tests/contrib/mue/test_seqs.fasta:1]: nit: please git remove this file and add `*.fasta` to `.gitignore`

File: tests/contrib/mue/test_statearrangers.py [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+                     -1/pf_arranger.epsilon*torch.ones([M, 3, 1])], axis=-1)
+    uln = torch.cat([torch.zeros([M, 3, 1]),
+                     -1/pf_arranger.epsilon*torch.ones([M, 3, 1])], axis=-1)
+    lln = torch.randn([D, B])
+    lln = lln - lln.logsumexp(-1, True)
+
+    a0ln, aln, eln = pf_arranger.forward(sln, cln, rln, uln, lln)
+
+    # --- Compute expected value per step. ---
+    # TODO: replace with VariableLengthDiscreteHMM function once implemented.
```
Comment (fritzo) [inline tests/contrib/mue/test_statearrangers.py:226]: is this TODO still relevant?

File: docs/source/contrib.mue.rst [hunk 16]
```diff
@@ -0,0 +1,44 @@
+MuE
+===
```
Comment (fritzo) [inline docs/source/contrib.mue.rst:2]: Could we make this more descriptive, say "Sequence alignment with MuE"?  This title appears in the docs.pyro.ai sidebar, so you'll it would be good to advertise it to people who don't already know what MuE means 😉

**Review by fritzo — COMMENTED** [review]
Hi @EWeinstein, this looks great.  I have only a few minor nits remaining.

Could we set up a time to quick walk over the code before merging, just so I can get an idea of future plans?

**Review by fritzo — APPROVED** [review]
Looks great!  I'll merge now and let's discuss your plans tomorrow.
